server/api/order.go

```diff
@@ -45,14 +58,119 @@ func CreateOrder(w http.ResponseWriter, r *http.Request) {
 	fmt.Fprint(w, payUrl)
 }
 
-func createOrder(order *model.Order) (string, error) {
-	id, url, err := payClient.Pay(order.Amount, "xxxx", "https://www.ec-cube.io/mall", "https://www.ec-cube.io/mall")
+func (api *API) ListOrder(w http.ResponseWriter, r *http.Request) {
+	starttime := r.URL.Query().Get("starttime")
+	endtime := r.URL.Query().Get("endtime")
+	id := r.URL.Query().Get("id")
+	name := r.URL.Query().Get("name")
+	phone := r.URL.Query().Get("phone")
+	status := r.URL.Query().Get("status")
+
+	filter := bson.D{{}}
+	if starttime != "" {
+		s, err := strconv.Atoi(starttime)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+		filter = append(filter, bson.E{"createtime", bson.D{{"$gt", s}}})
+	}
+	if endtime != "" {
+		s, err := strconv.Atoi(endtime)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+		filter = append(filter, bson.E{"createtime", bson.D{{"$lt", s}}})
+	}
+	if id != "" {
+		filter = append(filter, bson.E{"paymentid", id})
+	}
+	if phone != "" {
+		filter = append(filter, bson.E{"phone", phone})
+	}
+	if name != "" {
+		filter = append(filter, bson.E{"name", name})
+	}
+	if status != "" {
+		s, err := strconv.Atoi(status)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+		filter = append(filter, bson.E{"status", s})
+	}
+	orders, err := api.store.ListOrdersByFilter(filter)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	if len(orders) == 0 {
+		http.Error(w, "Not Found", http.StatusNotFound)
+		return
+	}
+	bs, _ := json.Marshal(orders)
+	w.WriteHeader(http.StatusOK)
+	w.Write(bs)
+}
+
+type UpdateReq struct {
+	ID     string
+	Status int
+}
+
+func (api *API) UpdateOrder(w http.ResponseWriter, r *http.Request) {
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, "Error reading request body", http.StatusInternalServerError)
+		return
+	}
+	var req UpdateReq
+	err = json.Unmarshal(body, &req)
+	if err != nil {
+		http.Error(w, "Error unmarshalling JSON", http.StatusBadRequest)
+		return
+	}
+	if req.ID == "" {
+		http.Error(w, "empty id", http.StatusBadRequest)
+		return
+	}
+	filter := bson.D{{"paymentid", req.ID}}
+	orders, err := api.store.ListOrdersByFilter(filter)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	if len(orders) == 0 {
+		http.Error(w, "Not Found", http.StatusNotFound)
+		return
+	}
+	if len(orders) > 1 {
+		http.Error(w, fmt.Sprintf("More than 1 order found for %s", req.ID), http.StatusInternalServerError)
+		return
+	}
+	order := orders[0]
+	order.Status = req.Status
+	if req.Status == model.StatusSent {
+		order.DepartureTime = int(time.Now().Unix())
+	}
+	if req.Status == model.StatusDeliverd {
+		order.FinishTime = int(time.Now().Unix())
```

Comment:
Before updating order, we might want to do order status validation here, i.e., comparing previous status with current and judge whether it is a reaonsble status transition. If not, there is bug somewhere (either in the backend or introduced by interaction among front end and backend)>

Reply:
Updated.